### ISSUE TITLE
Bug on IntentIntegrator prevents from using DecodeHints like ALLOWED_LENGTHS

### ISSUE THREAD

**Issue by silveriovale** [journeyapps/zxing-android-embedded#434]
#### Description of the problem:

Hello, guys.

There's a bug on IntentIntegrator that prevents users from using some DecodeHints, such as ALLOWED_LENGTHS.

The issue is on the "attachMoreExtras" method. It's used to map IntentIntegrator's extras to scanIntent extras. The problem is that it only maps correcly extras of the types Integer, Long, Boolean, Double, Float and Bundle. For values of other types, like ALLOWED_LENGTHS that is an int[], it transforms them into Strings.

```
private void attachMoreExtras(Intent intent) {
        for (Map.Entry<String, Object> entry : moreExtras.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();
            // Kind of hacky
            if (value instanceof Integer) {
                intent.putExtra(key, (Integer) value);
            } else if (value instanceof Long) {
                intent.putExtra(key, (Long) value);
            } else if (value instanceof Boolean) {
                intent.putExtra(key, (Boolean) value);
            } else if (value instanceof Double) {
                intent.putExtra(key, (Double) value);
            } else if (value instanceof Float) {
                intent.putExtra(key, (Float) value);
            } else if (value instanceof Bundle) {
                intent.putExtra(key, (Bundle) value);
            } else {
                intent.putExtra(key, value.toString());
            }
        }
    }
```

The problem is that ZXing's DecodeHintManager "parseDecodeHints(Intent intent)" validates the type of each hint (extra) and it filters out those hints whose value are not of the correct type (specified in DecodeHintType).

```
static Map<DecodeHintType, Object> parseDecodeHints(Intent intent) {
    Bundle extras = intent.getExtras();
    if (extras == null || extras.isEmpty()) {
      return null;
    }
    Map<DecodeHintType,Object> hints = new EnumMap<>(DecodeHintType.class);

    for (DecodeHintType hintType: DecodeHintType.values()) {

      if (hintType == DecodeHintType.CHARACTER_SET ||
          hintType == DecodeHintType.NEED_RESULT_POINT_CALLBACK ||
          hintType == DecodeHintType.POSSIBLE_FORMATS) {
        continue; // This hint is specified in another way
      }

      String hintName = hintType.name();
      if (extras.containsKey(hintName)) {
        if (hintType.getValueType().equals(Void.class)) {
          // Void hints are just flags: use the constant specified by the DecodeHintType
          hints.put(hintType, Boolean.TRUE);
        } else {
          Object hintData = extras.get(hintName);
          if (hintType.getValueType().isInstance(hintData)) {
            hints.put(hintType, hintData);
          } else {
            Log.w(TAG, "Ignoring hint " + hintType + " because it is not assignable from " + hintData);
          }
        }
      }
    }

    Log.i(TAG, "Hints from the Intent: " + hints);
    return hints;
  }
```

So, basically, your IntentIntegrator is changing the type of specific hints to String, and this is causing them to be discarded by DecodeHintManager.

I have been able to work around this issue by overriding your IntentIntegrator's "createScanIntent" method and adding the extra after the super method returns (after the wrong mapping occurs).

```
public class CustomIntentIntegrator extends IntentIntegrator {

    private int[] allowedLengths;

    public CustomIntentIntegrator(Activity activity) {
        super(activity);
    }

    public CustomIntentIntegrator(Activity activity, int[] allowedLengths) {
        super(activity);
        this.allowedLengths = allowedLengths;
    }

    @Override
    public Intent createScanIntent() {
        Intent intent = super.createScanIntent();
        intent.putExtra(DecodeHintType.ALLOWED_LENGTHS.name(), this.allowedLengths);
        return intent;
    }
}
```

Some of the DecodeHints can be very useful. For example: I need to scan ITFs of fixed lengths (44, 47 or 48) and if I'm not able to limit the lengths the scanner returns partial reads with lengths such as 30 or so. And if I try to filter valid results on the "onActivityResult", it has big performance issues - sometimes takes many seconds before continuing to scan. So by using ALLOWED_LENGTHS I'm able to greatly increase scanning peformance and provide a much greater user experience. (The scanner just keeps scanning, without lags or black screens, as it was ocurring when I was filtering on "onActivityResult".

By the way... C'mon guys... There's a "Kind of hacky" comment on the method...

``

**Which library version are you using?** E.g. 3.2.0.
3.6.0 or 3.3.0

Occurs on every device/emulator.

**Comment by rkistner** [comment]
Thanks for the detailed report. Will see if I can fix this in the next couple of days.

**Comment by rkistner** [comment]
Fixed by #469 in v4.0.2.
